### ISSUE TITLE
riverdatabasesql driver inserts empty unique key arrays instead of null

### ISSUE THREAD

**Issue by wydengyre** [riverqueue/river#650]
River migration 6 down fails in river 0.13.0, at least when using the `riverdatabasesql` driver, when there are periodic jobs in the job table. It fails with the following message:

    ERROR: could not create unique index "river_job_kind_unique_key_idx" (SQLSTATE 23505))

The relevant line from the migration is here: https://github.com/riverqueue/river/blob/56ddf095ab3b7ca183425641e84022a194d1111e/riverdriver/riverdatabasesql/migration/main/006_bulk_unique.down.sql#L11

It appears that the periodic jobs, rather than have their unique key set to `NULL`, instead just have an empty byte array (`\x`). As a result, they violate the unique constraint. Perhaps this was different in past versions of river?

**Comment by bgentry** [comment]
Dang, I think this ends up being a case where the migration isn't fully reversible if there's existing data conflicting with the index. 🤔 For performing the reverse migration, it's probably only going to be possible if you first clear or finish processing the existing unique jobs.

I'm not sure if there's anything that can be done to make this cleaner—really I wish we had just gotten this more flexible unique jobs implementation in the first place.

**Comment by brandur** [comment]
@bgentry Is this possibly indicative of a bug or at least unexpected behavior?  If we were to be storing an empty byte array, that probably means there wasn't supposed to be any unique considerations, and I would've expected that to a `NULL`.

**Comment by bgentry** [comment]
@brandur most likely yes, wish I had thought about that before. As I've verified this extensively on the pgx driver (just checked a long running live installation locally) it's probably something specific to dbsql.

**Comment by bgentry** [comment]
@brandur yep, confirmed it's a dbsql specific thing with the extra driver test coverage [here](https://github.com/riverqueue/river/compare/bg-dbsql-unique-bug). Working on figuring out the cause/fix.

**Comment by bgentry** [comment]
@brandur it seems to be that the `pq.Array` type being used for the `[][]byte` input for the multiple unique key byte slices isn't handling nil slices correctly, inserting an empty slice instead of a `NULL`. Haven't yet been able to figure out a fix, might be related to https://github.com/sqlc-dev/sqlc/issues/3021 or some other sqlc + pq issue.

**Comment by bgentry** [comment]
@wydengyre hi there, I just merged #739 which should resolve this. Would you mind giving it a try in your environment to see if it fixes your issues with unique jobs in riverdatabasesql?

**Comment by wydengyre** [comment]
@bgentry I appear to now be getting errors like this on insertion of unique jobs:

`ERROR: invalid byte sequence for encoding "UTF8": 0xf4 0x8d 0x61 0x24 (SQLSTATE 22021)`

**Comment by bgentry** [comment]
@wydengyre can you provide some details about your environment? Are you using plain Postgres, which version, is pgbouncer involved, any specific database/sql driver involved, etc. I'm having trouble reproducing this locally—even the extra failing test coverage I linked to before works locally on the latest commit.

**Comment by bgentry** [comment]
@wydengyre ahh, j/k, I was able to reproduce! Just needed to try it with a byte string that's not valid UTF8. Will open a new issue shortly because it's a different cause than this one.
